### PR TITLE
More extensive examples of graph building.

### PR DESCRIPTION
**Pull request by diyessi**

### CODE REVIEW

File: doc/examples/abc_operator.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    // Get the backend
+    auto manager = runtime::Manager::get("CPU");
+    auto backend = manager->allocate_backend();
+
+    // Compile the function
+    auto external = manager->compile(f);
+    auto cf = backend->make_call_frame(external);
+
+    // Allocate tensors
```
Comment (aprocter) [inline doc/examples/abc_operator.cpp]: Might specify that these are for args and results specifically.

File: doc/examples/mnist.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
+/*******************************************************************************
```
Comment (aprocter) [inline doc/examples/mnist.cpp]: Might consider renaming this file and its associated `.hpp` to `mnist_loader` or something. For whatever reason, at first glance I was expecting to find an implementation of a model in here.

File: doc/examples/mnist_mlp.cpp [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+                      const std::shared_ptr<runtime::TensorView>& t_Y)
+{
+    const Shape& sm_shape = t_sm->get_shape();
+    size_t batch_size = sm_shape.at(0);
+    size_t label_size = sm_shape.at(1);
+    const Shape& Y_shape = t_Y->get_shape();
+    if (Y_shape.size() != 1 || Y_shape.at(0) != batch_size)
+    {
+        throw std::invalid_argument(
+            "Y and softmax shapes are incomptible");
```
Comment (aprocter) [inline doc/examples/mnist_mlp.cpp]: Typo (`incomptible`)

File: doc/examples/mnist_mlp.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+#include "mnist.hpp"
+#include "tensor_utils.hpp"
+
+using namespace ngraph;
+
+size_t accuracy_count(const std::shared_ptr<runtime::TensorView>& t_sm,
+                      const std::shared_ptr<runtime::TensorView>& t_Y)
+{
+    const Shape& sm_shape = t_sm->get_shape();
```
Comment (aprocter) [inline doc/examples/mnist_mlp.cpp]: `t_sm` is a bit cryptic---might spell it out as `softmax`

File: doc/examples/mnist_mlp.cpp [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    size_t batch_size = sm_shape.at(0);
+    size_t label_size = sm_shape.at(1);
```
Comment (aprocter) [inline doc/examples/mnist_mlp.cpp]: Maybe `label_count`?

File: doc/examples/mnist_mlp.cpp [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    const Shape& Y_shape = t_Y->get_shape();
+    if (Y_shape.size() != 1 || Y_shape.at(0) != batch_size)
+    {
+        throw std::invalid_argument(
+            "Y and softmax shapes are incomptible");
+    }
+    size_t sm_pos = 0;
+    size_t count = 0;
+    for (size_t i = 0; i < batch_size; ++i)
```
Comment (aprocter) [inline doc/examples/mnist_mlp.cpp]: The loop structure here is a bit subtle. Could we somehow cast the data to a 2D array?

File: doc/examples/mnist_mlp.cpp [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+                   0,
+                   train_loader.get_image_batch_size() * sizeof(float));
+        t_Y->write(train_loader.get_label_floats(),
+                   0,
+                   train_loader.get_label_batch_size() * sizeof(float));
+        train_cf->call(
+            {t_loss, t_sm, t_W0_next, t_b0_next, t_W1_next, t_b1_next},
+            {t_X, t_Y, t_N, t_learning_rate, t_W0, t_b0, t_W1, t_b1});
+
+        t_W0.swap(t_W0_next);
```
Comment (aprocter) [inline doc/examples/mnist_mlp.cpp]: nice :)

File: doc/examples/update.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
```
Comment (aprocter) [inline doc/examples/update.cpp]: What is this file for?

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+Documentation in this section describes one of the ways to derive a trainable 
+model from an inference model.
+
+We can derive a trainable model from any graph that has been constructed with 
+weight-based updates. For this example named ``mnist_mlp.cpp``, we start with a hand-designed inference model and convert it to a model that can be trained 
+with nGraph. 
+
+Additionally, and to provide a more complete walk-through that *also* trains the 
+"trainable" model, our example includes the use of a simple data loader for the
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: I don't think we need the quotation marks here.

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  - :ref:`update`
+
+
+.. _model_overview:
+
+Model overview
+==============
+
+The nGraph API was designed for automatic graph construction under direction of 
+a framework. Without a framework, the process is somewhat tedious, so the example
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: Maybe we should be a bit less self deprecating here. :)  Something like "The graph construction API is lower-level than a typical framework, so writing a model in it directly is somewhat akin to programming in assembly language. Because of this, the example selected is..."

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+Model overview
+==============
+
+The nGraph API was designed for automatic graph construction under direction of 
+a framework. Without a framework, the process is somewhat tedious, so the example 
+selected is a relatively simple model: a fully-connected topology with one hidden 
+layer followed by ``Softmax``.
+
+Since the graph is stateless, parameters for the input and the variables must be
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: I don't quite understand this sentence. (What are "original" weights?)

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+   :language: cpp
+   :lines: 150-164
+
+
+.. _backprop:
+
+Backprop
+--------
+
+To compute the updates, we need a computation that computes an
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: maybe s/`the updates`/`weight updates`/ (or `parameter updates`)

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+adjustment for the weights from an adjustment to the loss. In nGraph,
+``loss`` is the computation that computes the loss; it is equivalent
+to what some descriptions of the autodiff algorithm call "the tape."
+If each step of a computation between a weight and the loss has a
+derivative, we can use the reverse mode autodiff to compute an update
+for the weight from an update for the loss; in fact, we can compute
+updates for all the weights, sharing much of the update computation,
+and this is what some frameworks do. But it is just as easy for us to
+instead create the update computations for all of the weights, which
+lets compilation optimize across steps in the computation.
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: I think the distinction being drawn here is between Brand X, in which each node "knows how to do its own backprop", and nGraph where we have something that automatically constructs a graph for the backprop?

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+We'll call the adjustment to the loss
+
+.. code-block:: cpp
+
+   auto delta = -learning_rate * loss;
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: should this be a `literalinclude`?

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+For any node ``N``, if the update for ``loss`` is ``delta``, the
+update computation for ``N`` will be given by the node
+
+.. code-block:: cpp
+
+   auto update = loss->backprop_node(N, delta);
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: `literalinclude`?

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+want to track how well we are doing; this is a function that returns
+the loss and has the labeled testing data as input. Although we can
+use the same nodes in different functions, nGraph currently does not
+allow the same nodes to be compiled in different functions, so we
+compile clones of the nodes.
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 246-253
+
```
Comment (aprocter) [inline doc/sphinx/source/howto/derive-for-training.rst]: Some example output might be good at the end here. Also, are we giving people the information they need to prepare the dataset? (Maybe that's in a different file)

File: doc/sphinx/source/howto/update.rst [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+In this section, we show how to make a stateful computation from
+nGraphs stateless operations. The basic idea is that any computation
+with side-effects can be factored into a stateless function that
+transforms the old state into the new state.
+
+An Example from C++
+===================
+
+Let's start with a simple C++ example, a function ``count`` that
```
Comment (aprocter) [inline doc/sphinx/source/howto/update.rst]: Ahh, now I see what `update.cpp` was for  :)

File: doc/sphinx/source/howto/derive-for-training.rst (hunk 15 above)
Comment (indie) [reply doc/sphinx/source/howto/derive-for-training.rst]: This line of code is not part of the original example for how to build, right?  A `literalinclude::` would not make sense here.  This is more like a demonstrative part of the "how" the auto-updater updates?

File: doc/sphinx/source/howto/derive-for-training.rst [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+.. code-block:: cpp
+
+   auto update = loss->backprop_node(N, delta);
+
+The different update nodes will share intermediate computations. So to
+get the updated value for the weights we just say:
+
+.. literalinclude:: ../../../examples/mnist_mlp.cpp
+   :language: cpp
+   :lines: 166-176
```
Comment (indie) [inline doc/sphinx/source/howto/derive-for-training.rst]: lines 172-176?

File: doc/sphinx/source/howto/update.rst [hunk 19]
```diff
@@ -0,0 +1,48 @@
+.. update.rst
+
+###########################
+Make a Stateful Computation
```
Comment (indie) [inline doc/sphinx/source/howto/update.rst]: Sentence case

File: doc/sphinx/source/howto/operator.rst [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+############################
+Build a graph with operators
+############################
+
+This section illustrates the use of C++ operators to simplify the
+building of graphs.
+
+Several C++ operators have been overloaded so that when given nodes as
+arguments they construct a node that performs the operation on the
```
Comment (indie) [inline doc/sphinx/source/howto/operator.rst]: Several C++ operators have been overloaded such that when given nodes as arguments, they may construct a simplified node to perform the operation / that performs the operation ...

File: doc/sphinx/source/howto/update.rst [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+###########################
+
+In this section, we show how to make a stateful computation from
+nGraphs stateless operations. The basic idea is that any computation
```
Comment (indie) [inline doc/sphinx/source/howto/update.rst]: Note:  My guess is that this kind of picks up from where @Krovatkin and I left off working on documentation around op fusion.  How does this intro sentence from [here](https://github.com/NervanaSystems/ngraph/blob/leona/21_Mar/doc/sphinx/source/howto/fuse.rst) work?  

> In working with a function graph, there are many ways to describe what happens when we need to do something with the ops (AKA "operational functions") from that graph.   

Then... 

> This section describes how to make a *stateful* computation from the *statelessness* of nGraph ops.  

?

File: doc/sphinx/source/howto/update.rst [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+===================
+
+Let's start with a simple C++ example, a function ``count`` that
+returns how many times it has already been called:
+
+.. literalinclude:: ../../../examples/update.cpp
+   :language: cpp
+   :lines: 20-24
+
+This function has state kept in the static variable ``counter``. The
```
Comment (indie) [inline doc/sphinx/source/howto/update.rst]: This intro might need reworded or edited.  
Could this part 
> this function has state kept in the static variable ``counter``

Be reworded for clarity to 

> A static variable named ``counter`` holds the state of this function (... which is initialized to zero) 

?

File: doc/sphinx/source/howto/derive-for-training.rst (hunk 15 above)
Comment (aprocter) [reply doc/sphinx/source/howto/derive-for-training.rst]: > This line of code is not part of the original example for how to build, right?

My bad, I assumed it was.

File: doc/sphinx/source/howto/update.rst (hunk 21 above)
Comment (diyessi) [reply doc/sphinx/source/howto/update.rst]: They are different things.
